### PR TITLE
scratchpad support, multiregion enclaves, etc

### PR DESCRIPTION
**Pull request by dkohlbre**
FU540 plugin changes to add scratchpad support, fix cache part bugs, etc.
Added a generic multi-region interface for enclaves to allow for  >2 PMP logical regions.
Plugin for SBI to talk to these regions.

### CODE REVIEW

**Comment by dayeol** [comment]
Will the on-chip enclave be enabled by default on FU540?

**Comment by dkohlbre** [comment]
No.
We need to get a interface for asking for specific mitigations working first in my opinion.
Cache partitioning and on-chip execution should be off by default for now.

File: sm/platform/fu540/fu540_internal.c [hunk 1]
```diff
@@ -6,40 +6,210 @@
...
+  }
+
+  /* At this point, no master has waymasks for the scratchpad ways,
+     and all scratchpad addresses have L2 lines */
+
+  /* We try and check it now, any error SHOULD be immediately detectable. */
+  for(addr = scratch_start; addr < scratch_stop; addr += L2_LINE_SIZE){
+    if(*(uintptr_t*)addr != 64){
+      printm("FATAL: Found a bad line %x\r\n", addr);
+      /* TODO Not fatal! */
```
Comment (dkohlbre) [inline sm/platform/fu540/fu540_internal.c]: make it fatal

File: sm/platform/fu540/fu540_internal.c [hunk 2]
```diff
@@ -6,40 +6,210 @@
...
+
+  waymask_init();
+  scratchpad_allocated_ways = 0;
+
+  /* PMP Lock the entire L2 controller */
+  if(pmp_region_init_atomic(CACHE_CONTROLLER_ADDR_START,
+                            CACHE_CONTROLLER_ADDR_END - CACHE_CONTROLLER_ADDR_START,
+                            PMP_PRI_ANY, &l2_controller_rid, 1)){
+    printm("FATAL CANNOT CREATE PMP FOR CONTROLLER\r\n");
+    //TODO: this isn't currently fatal!
```
Comment (dkohlbre) [inline sm/platform/fu540/fu540_internal.c]: make it fatal

File: sm/platform/fu540/fu540_internal.c [hunk 3]
```diff
@@ -6,40 +6,210 @@
...
+  }
+  /* Create PMP region for scratchpad */
+  if(pmp_region_init_atomic(L2_SCRATCH_START,
+                            L2_SCRATCH_STOP - L2_SCRATCH_START,
+                            PMP_PRI_ANY, &scratch_rid, 1)){
+    printm("FATAL CANNOT CREATE SCRATCH PMP\r\n");
+    //TODO: this isn't currently fatal!
```
Comment (dkohlbre) [inline sm/platform/fu540/fu540_internal.c]: make it fatal

File: sm/platform/fu540/fu540_internal.c [hunk 4]
```diff
@@ -6,40 +6,210 @@
...
+
+    /* Copy the enclave over */
+    uintptr_t old_epm_start = pmp_region_get_addr(enclave->regions[old_epm_idx].pmp_rid);
+    uintptr_t scratch_epm_start = pmp_region_get_addr(scratch_rid);
+    size_t size = enclave->pa_params.free_base - old_epm_start;
+    size_t scratch_size = 8*L2_WAY_SIZE;
+
+    //TODO need a check and failure mode here
+    if(size > scratch_size){
+      printm("FATAL: Enclave too big for scratchpad!\r\n");
```
Comment (dkohlbre) [inline sm/platform/fu540/fu540_internal.c:144]: make it fatal

File: sm/platform/fu540/fu540_internal.c [hunk 5]
```diff
@@ -6,40 +6,210 @@
...
 
 }
 
-void platform_switch_from_enclave(struct platform_enclave_data_t* ped){
-  /* Free all our ways */
-  waymask_free_ways(ped->saved_mask);
+void platform_destroy_enclave(struct enclave* enclave){
+  if(enclave->ped.use_scratch){
+    /* Clean out the region ourselves */
+    /* TODO wipe */
```
Comment (dkohlbre) [inline sm/platform/fu540/fu540_internal.c]: clean up the scratchpad memory

File: sm/sm.mk.in [hunk 6]
```diff
@@ -18,6 +18,8 @@ sm_c_srcs = \
   ed25519/ge.c \
   ed25519/sc.c \
   platform/@TARGET_PLATFORM@/@TARGET_PLATFORM@.c \
+	plugins/plugins.c \
+# TODO: should be @PLUGIN_SOURSES@ \
```
Comment (dkohlbre) [inline sm/sm.mk.in]: can we fix this to be properly generic?

**Review by dkohlbre — COMMENTED** [review]
A bunch of small todos/etc should get cleaned up.
@dayeol can we fix this new plugins source includes?

**Comment by dayeol** [comment]
@dkohlbre Please keep the commits when you merge this PR. There're multiple changes and way too many commits, so squashing will make it a little hard to track down the history in my opinion.

File: sm/sm.mk.in (hunk 6 above)
Comment (dayeol) [reply sm/sm.mk.in]: I made it a flag `--enable-sm-multimem`

File: sm/platform/fu540/fu540_internal.c (hunk 5 above)
Comment (dayeol) [reply sm/platform/fu540/fu540_internal.c]: is this resolved?

**Pull request merged**